# Differential methylation analysis using bedmethyl files from long-read (ONT) data
***`ont-methylDMR-kit`*** is a pipeline to call differentially methylated regions (DMRs) between `haplotypes`, `two samples`, or between `two groups`, coupled with annotation and visualisation, on long-read (ONT) sequencing bedmethyl files generated using [modkit](https://github.com/nanoporetech/modkit). This pipeline is inspired by my work on rare disorders, and the fact that long-read sequencing has the potential to comprehensively identify all modified bases, such as 5-Methylcytosine (5mC), 5-Hydroxymethylcytosine (5hmC), N6-methyladenine (6mA), and N4-methylcytosine (4mC) that have been identified for a growing number of rare disorders and imprinted disorders.

The pipeline is built using [Nextflow](https://www.nextflow.io/), a bioinformatics workflow manager that enables the development of portable and reproducible workflows. 
There are three docker images available from [DockerHub](https://hub.docker.com/repository/docker/nyagam/ont-methyl-kit/general); [DockerHub](https://hub.docker.com/repository/docker/nyagam/modbamtools-v0.4.8/general); and [DockerHub](https://hub.docker.com/repository/docker/nyagam/methylartist/tags) that contains all the tools/softwares required by the pipeline, making results highly reproducible. 

# DMR analysis:
DMR analysis is performed using the R-package [DSS](https://bioconductor.org/packages/release/bioc/vignettes/DSS/inst/doc/DSS.html). It supports calling DMRs across 5-Methylcytosine (`--5mC`), 5-Hydroxymethylcytosine (`--5hmC`), N6-methyladenine (`--6mA`), and N4-methylcytosine (`--4mC`) modified bases. It also supports haplotype-specific DMRs using `--phased_mC`, `--phased_mA`, and `--phased_hmC`. Just provide raw bedmethyl files as generated by [modkit](https://github.com/nanoporetech/modkit) using either `--input_file1` and `--input_file2` for calling DMRs between haplotypes or two samples; or `--input_group1` and `--input_group2` for group analysis (bedmethyls in these two folders must have `*.bed` extension). Note that `--phased_mC` and
`--phased_hmC` flags are not supported in group analysis; use either `--5mC`, `--4mC` or `--6mA` separately depending on the type of methylation being analysed. Methyl positions with less than 5 reads are filtered by default. The current default options for [DSS](https://bioconductor.org/packages/release/bioc/vignettes/DSS/inst/doc/DSS.html); delta (threshold for defining DMR) at 10%, p-values threshold for calling DMR at 0.01, minimum length (in basepairs) required for DMR methylation change analysis at 100, minimum number of CpG sites required for DMR at 10, and merging two DMRs that are very close to each other at less than 100 basepairs. To change these parameters, edit `main.nf` in `process dmr_calling or process group_dmr_calling_5mC or process group_dmr_calling_6mA` See [DSS](https://bioconductor.org/packages/release/bioc/vignettes/DSS/inst/doc/DSS.html) for more information. 

# Annotation:
Significant DMRs are annotated to provide information on whether DMRs overlap with promoters, exons, and introns. A compressed file, annotations.zip (which needs to be unzipped `tar -xvf annotations.zip`), is provided with the pipeline that contains the annotation information, which is based on gencode v46. Users can also annotate and plot DMR regions overlapping imprinted genes (provided wiht the pipeline), using `--imprinted`. Imprinted gene list constist of 235 genes based on varying levels of imprinting evidence and associated disorders, and is derived from [PanelApp UK](https://panelapp.genomicsengland.co.uk/panels/227/); and [PanelApp Australia](https://panelapp-aus.org/panels/3663/).

# Visualisation:
Annotated DMRs are plotted using [modbamtools](https://github.com/rrazaghi/modbamtools) and [methylartist](https://github.com/adamewing/methylartist). It supports haplotype-specific DMR plotting (by providing haplotagged modified bam files using `--phased_modBam` ) or DMRs between two samples (by providing modified bam files using `--input_modBam1` and `--input_modBam2`) but does not support plotting DMRs from group analysis. You can provide a gene list with `--gene_list` to only plot significant DMRs for the provided genes (if present), or `--imprinted` to plot significant DMRs overlapping imprinted genes. To plot DMRs using `--methylartist_only`, `--reference` must be provided. 

There is a plots only mode triggered by the flag `--plots_only`, which requires an annotated DMR bed file `--annotated_dmrs` (generated by `process annotate_dmrs` ), modified bam files, and either `--phased_mC`, `--phased_hmC` for phased; or `--5mC`, `--4mC` for two-sample analysis. 

# Installation and Usage:
```bash
$ git clone https://github.com/NyagaM/ont-methylDMR-kit.git
$ cd ont-methylDMR-kit
$ tar -xvf annotations.tar.gz
```
To view usage and run options:

```bash
$ nextflow run main.nf --help
```
```bash
Usage: nextflow run main.nf [options]

Options:
  --input_file1        First input bedmethyl file from sample_1 or haplotype_1 bedmethyl (required for analysis)
  --input_file2        Second input bedmethyl file from sample_2 or haplotype_2 bedmethyl (required for analysis)
  --output_dir         Output directory (required)
  --input_modBam1      First input modified BAM used to generate bedmethyl for sample_1 (optional, for plotting)
  --input_modBam2      Second input modified BAM used to generate bedmethyl for sample_2 (optional, for plotting)
  --input_group1       Directory containing bedmethyl files (with *.bed extension) for group 1 (optional, for group analysis)
  --input_group2       Directory containing bedmethyl files (with *.bed extension) for group 2 (optional, for group analysis)
  --gene_list          A list of genes to generate plots on (optional). If not provided and --imprinted is used, imprinted genes will be plotted.
  --reference          Reference genome FASTA file (required for methylartist plots)
  --imprinted          Use this flag to plot DMRs overlapping imprinted genes if --gene_list is not provided (optional, for haplotagged plots)
  --5mC                Use this flag to trigger 5mC DMR calling (optional)
  --5hmC               Use this flag to trigger 5hmC DMR calling (optional)
  --6mA                Use this flag to trigger 6mA DMR calling (optional)
  --4mC                Use this flag to trigger 4mC DMR calling (optional)
  --phased_mC          Use this flag to trigger haplotagged 5mC/4mC DMR calling if input files are haplotagged bedmethyls (optional)
  --phased_mA          Use this flag to trigger haplotagged 6mA DMR calling if input files are haplotagged bedmethyls (optional)
  --phased_hmC         Use this flag to trigger haplotagged 5hmC DMR calling if input files are haplotagged bedmethyls (optional)
  --phased_modBam      Haplotagged modified BAM (required for plotting DMRs if --phased_mC, --phased_mA or --phased_hmC is used)
  --plots_only         Only run the plotting processes, requires --annotated_dmrs and relevant BAM files (and --reference for methylartist)
  --annotated_dmrs     Path to annotated DMR bed file (required if --plots_only is used)
  --methylartist_only  Use this flag to only generate plots using methylartist (requires --reference). Otherwise, both modbamtools and methylartist plots are generated if reference is provided.
  --help               Print this help message
```


Users can run the test data provided to test installation and set-up of the pipeline as follows:
```
cd ont-methylDMR-kit && tar -xvf test_data.tar.gz

nextflow run ont-methylDMR-kit/main.nf -profile standard \
  --input_file1 ont-methylDMR-kit/HG002_base-mod-5mC_chr15/HG002_chr15_5mC.1.bed \
  --input_file2 ont-methylDMR-kit/HG002_base-mod-5mC_chr15/HG002_chr15_5mC.2.bed \
  --phased_mC \
  --phased_modBam ont-methylDMR-kit/HG002_base-mod-5mC_chr15/HG002_base-mod-5mC_chr15.bam \
  --output_dir output \
  --imprinted
  ##--reference provide a reference to plot DMRs using methylartist as well
```
The output dir should have the following: 

`annotated_dmrs`: folder containing annotated dmrs and log file
`dmrs`: folder with dmrs and debug folder containing R-scripts and log files
`haplotagged_dmr_plots`: folder with plots and log file
`imprinted_genes.txt`: imprinted gene list
`modified_beds`: folder with modified input bed files. 

Below is an examples of a significant 2.4kb haplotype-specific DMR identified using the pipeline within the imprinted *SNURF* gene. *SNURF* is mapped within the Prader-Willi Syndrome critical region on chromosome 15. The transcripts produced from this gene initiate at an imprinting center and are paternally-imprinted.

**Visualisation of the DMR using [modbamtools](https://github.com/rrazaghi/modbamtools)**

![newplot (7)](https://github.com/user-attachments/assets/3464d66b-3c8f-44df-89f6-efb2a770cef0)


**Visualisation of the DMR using [methylartist](https://github.com/adamewing/methylartist)**

![chr15_24954397-24956828_SNURF](https://github.com/user-attachments/assets/455ad8f2-e913-4cc4-9221-00e615bceee6)



To run the full DMR analysis workflow between between two samples:
```
nextflow run ont-methylDMR-kit/main.nf -profile standard \
  --input_file1 /path/to/bedmethyl file for sample 1 \
  --input_file2 /path/to/bedmethyl file for sample 2 \
  --5mC \ # or --6mA or --4mC
  --input_modbam1 /path/to/modBam for sample 1 \
  --input_modbam2 /path/to/modBam for sample 2 \
  --output_dir /path/to/write output \
  --gene_list /path/to/gene_list.txt  # if not provided, all regions will be plotted or use --imprinted to plot across imprinted genes
```
To run plots-only mode for DMRs identified between two samples:
```
nextflow run ont-methylDMR-kit/main.nf -profile standard \
  --plots-only \
  --annotated_dmrs /path/to/dmrs_table_annotated.bed \
  --input_modbam1 /path/to/modBam for sample 1 \
  --input_modbam2 /path/to/modBam for sample 2 \
  --output_dir /path/to/write output \
  --gene_list /path/to/gene_list.txt or --imprinted
```
To run plots-only mode for DMRs identified between haplotypes:
```
nextflow run ont-methylDMR-kit/main.nf -profile standard \
  --plots-only \
  --phased_mC \ # or --phased_mA or --phased_hmC
  --annotated_dmrs /path/to/dmrs_table_annotated.bed \
  --phased_modBam /path/to/phased modBam for the sample \
  --output_dir /path/to/write output \
  --gene_list /path/to/gene_list.txt or --imprinted
```
To run DMR analysis between two groups of bedmethyl files:
```
nextflow run ont-methylDMR-kit/main.nf -profile standard \
  --input_group1 /path/to/bedmethyl files (must have .bed extension) for group 1 \
  --input_group2 /path/to/bedmethyl files (must have .bed extension) for group 2 \
  --5mC \ # or --6mA or --4mC 
  --output_dir /path/to/write output
```
# Publications
1. Nyaga, D.M., Tsai, P., Gebbie, C. et al. Benchmarking nanopore sequencing and rapid genomics feasibility: validation at a quaternary hospital in New Zealand. npj Genom. Med. 9, 57 (2024). https://doi.org/10.1038/s41525-024-00445-5
